pkg/pipeline/pipeline_builder.go

```diff
@@ -0,0 +1,335 @@
+		dstEntry, ok := b.pipelineEntryMap[connection.Name]
+		if !ok {
+			return nil, fmt.Errorf("unknown pipeline stage: %s", connection.Name)
+		}
+		dstNode, err := b.getStageNode(dstEntry)
+		if err != nil {
+			return nil, err
+		}
+		dst, ok := dstNode.(node.Receiver)
+		if !ok {
+			return nil, fmt.Errorf("stage %q of type %q can't receive data",
+				connection.Name, dstEntry.stageType)
+		}
+		// instantiates (or loads from cache) the source node of a connection
+		srcEntry, ok := b.pipelineEntryMap[connection.Follows]
+		if !ok {
+			return nil, fmt.Errorf("unknown pipeline stage: %s", connection.Follows)
+		}
+		srcNode, err := b.getStageNode(srcEntry)
+		if err != nil {
+			return nil, err
+		}
+		src, ok := srcNode.(node.Sender)
+		if !ok {
+			return nil, fmt.Errorf("stage %q of type %q can't send data",
+				connection.Name, dstEntry.stageType)
+		}
+		log.Debugf("connecting stages: %s --> %s", connection.Follows, connection.Name)
+
+		// connects source and destination node, and catches any panic from the Go-Pipes library.
+		var catchErr error
+		func() {
+			defer func() {
+				if msg := recover(); msg != nil {
+					catchErr = fmt.Errorf("%q and %q stages haven't compatible input/outputs: %v",
+						connection.Follows, connection.Name, msg)
+				}
+			}()
+			src.SendsTo(dst)
+		}()
+		if catchErr != nil {
+			return nil, catchErr
+		}
+	}
+
+	if len(b.startNodes) == 0 {
+		return nil, errors.New("no ingesters have been defined")
+	}
+	if len(b.terminalNodes) == 0 {
+		return nil, errors.New("no writers have been defined")
+	}
+	return &Pipeline{
+		startNodes:     b.startNodes,
+		terminalNodes:  b.terminalNodes,
+		pipelineStages: b.pipelineStages,
+	}, nil
+}
+
+func (b *builder) getStageNode(pe *pipelineEntry) (interface{}, error) {
+	if stg, ok := b.createdStages[pe.configStage.Name]; ok {
+		return stg, nil
+	}
+	var stage interface{}
+	// TODO: modify all the types' interfaces to not need to write loops here, the same
+	// as we do with Ingest
+	switch pe.stageType {
+	case StageIngest:
+		init := node.AsInit(pe.Ingester.Ingest)
+		b.startNodes = append(b.startNodes, init)
+		stage = init
+	case StageWrite:
+		term := node.AsTerminal(func(in <-chan []config.GenericMap) {
+			for i := range in {
+				pe.Writer.Write(i)
+			}
+		})
+		b.terminalNodes = append(b.terminalNodes, term)
+		stage = term
+	case StageDecode:
+		stage = node.AsMiddle(func(in <-chan []interface{}, out chan<- []config.GenericMap) {
+			for i := range in {
+				out <- pe.Decoder.Decode(i)
+			}
+		})
+	case StageEncode:
+		stage = node.AsMiddle(func(in <-chan []config.GenericMap, out chan<- []config.GenericMap) {
+			for i := range in {
+				out <- pe.Encoder.Encode(i)
+			}
+		})
+	case StageTransform:
+		stage = node.AsMiddle(func(in <-chan []config.GenericMap, out chan<- []config.GenericMap) {
+			for i := range in {
+				out <- transform.ExecuteTransform(pe.Transformer, i)
+			}
```

Comment:
`transform` could now be the same structure as `encode` or `extract`, so we can do away with the `ExecuteTransform()` intermediary and call `pe.Transformer.Transform()` directly.

Reply:
Currently, this would involve adding the loop inside Execute transform in any transform implementation (as the `Transform` interface accepts and returns a single `GenericMap` and here we are playing with arrays).

I was thinking on address your proposal in a later PR, where we put in common all the functional interfaces and change them by input/output channels.